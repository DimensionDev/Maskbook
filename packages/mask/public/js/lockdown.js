/// <reference types="ses" />
if (
    Object.isExtensible(Array.prototype) &&
    // https://github.com/facebook/react/issues/27005
    (typeof location === 'object'
        ? location.protocol.endsWith('extension:') && location.pathname !== '/devtools-background.html'
        : true)
) {
    try {
        // lockdown might fail when there is new well known Symbol
        lockdown({
            // In production, we have CSP enforced no eval
            // In development, we use Trusted Types.
            evalTaming: 'unsafeEval',
            //
            overrideTaming: 'severe',
            consoleTaming: 'unsafe',
            errorTaming: 'unsafe',
            errorTrapping: 'none',
            unhandledRejectionTrapping: 'none',
        })
    } catch {
    } finally {
        // https://github.com/endojs/endo/issues/1345 and https://github.com/endojs/endo/issues/1346
        console.clear()
    }
}

undefined
