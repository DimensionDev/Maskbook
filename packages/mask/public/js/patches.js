// this key is also used for bundled React Devtools in Mask Network
if (!globalThis[Symbol.for('mask_init_patch')]) {
    globalThis[Symbol.for('mask_init_patch')] = true
    // Fix for globalThis !== window in content script in Firefox
    patch1: try {
        /**
         * Many methods on `window` requires `this` points to a Window object
         * Like `alert()`. If you call alert as `const w = { alert }; w.alert()`,
         * there will be an Illegal invocation.
         *
         * To prevent `this` binding lost, we need to rebind it.
         *
         * @param desc {PropertyDescriptor | undefined}
         * @param global {Window}
         */
        const PatchThisOfDescriptorToGlobal = (desc, global) => {
            if (!desc) return
            const { get, set, value } = desc
            if (get)
                desc.get = function () {
                    if (this === globalThis) return get.apply(window)
                    return get.apply(this)
                }
            if (set)
                desc.set = function (val) {
                    if (this === globalThis) return set.apply(global, val)
                    return set.apply(this, val)
                }
            if (value && typeof value === 'function') {
                const desc2 = Object.getOwnPropertyDescriptors(value)
                desc.value = function () {
                    if (new.target) return Reflect.construct(value, arguments, new.target)
                    return Reflect.apply(value, this === globalThis ? global : this, arguments)
                }
                Object.defineProperties(desc.value, desc2)
                try {
                    desc.value.prototype = value.prototype
                } catch {}
            }
        }
        // MV3 service worker
        if (typeof window === 'undefined') break patch1
        // In the content script, globalThis !== window.
        if (globalThis === window || !('browser' in globalThis)) break patch1

        // See: https://bugzilla.mozilla.org/show_bug.cgi?id=1577400
        /**
         * @type {(keyof typeof window & keyof typeof globalThis)[]}
         * If any API is complaining "this does not implements Window", add it here.
         */
        const brokenAPI = [
            'requestAnimationFrame',
            'setTimeout',
            'clearTimeout',
            'matchMedia',
            'getComputedStyle',
            'getSelection',
            'requestIdleCallback',
        ]
        const patch = Object.create(null)
        for (const api of brokenAPI) {
            const desc = Object.getOwnPropertyDescriptor(window, api)
            patch[api] = desc
            PatchThisOfDescriptorToGlobal(desc, window)
        }
        console.log('[Mask] applying intrinsic patches', patch)
        Object.defineProperties(window, patch)
    } catch {}

    patch2: try {
        if (location.protocol !== 'safari-web-extension:') break patch2
        /**
         * @param {Response | PromiseLike<Response>} source
         */
        WebAssembly.compileStreaming = async function (source) {
            const response = await source
            if (response.headers.get('content-type') === 'application/wasm') {
                console.warn(`[Mask] Safari WebAssembly.compileStreaming patch is no longer needed.`)
            }
            const buffer = await response.arrayBuffer()
            return WebAssembly.compile(buffer)
        }
        /**
         *
         * @param {Response | PromiseLike<Response>} source
         * @param {WebAssembly.Imports | undefined} importObject=
         */
        WebAssembly.instantiateStreaming = async function (source, importObject) {
            const response = await source
            if (response.headers.get('content-type') === 'application/wasm') {
                console.warn(`[Mask] Safari WebAssembly.instantiateStreaming patch is no longer needed.`)
            }
            const buffer = await response.arrayBuffer()
            return WebAssembly.instantiate(buffer, importObject)
        }
    } catch {}
}

undefined
