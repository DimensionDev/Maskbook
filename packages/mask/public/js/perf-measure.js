// @ts-check
// this file is only used for module performance analyze.
;(() => {
    const regex = /^\.\/node_modules\/\.pnpm\/(?<key>[^_\/]+)/
    const npm_name_cache = Object.create(null)
    const readable_name_cache = Object.create(null)
    const IMPORT_STYLE = `color: purple;`
    const USED_BY_STYLE = `color: green;`
    const FAST_TIME_STYLE = `color: gray;`
    const MERGED_MODULE_STYLE = `color: gray;`
    const SLOW_TIME_STYLE = `color: red;`
    const TIME_STYLE = `color: brown;`
    const EMPTY = Object.freeze([])

    class Graph {
        /** @type {readonly Module[]} */
        modules
        /** @type {readonly Module[]} */
        root_or_async_or_deferred_module
        total_time = 0
        constructor() {
            const observer = new PerformanceObserver((entryList) => {
                this.#freeze()
                console.clear()

                const [entry] = entryList.getEntries()
                console.log('LCP result:', entry)
                console.log('Type measure to see the graph.\n', this)
                console.log(
                    '- Modules:',
                    this.#modules.size,
                    '\n- First module:',
                    this.#first_request,
                    '\n- Total time:',
                    this.total_time,
                    '\n- Largest contentful paint:',
                    entry.startTime,
                )
                observer.disconnect()
            })
            observer.observe({ type: 'largest-contentful-paint', buffered: false })
        }
        #finalized = false
        /** @type {Module[]} */
        #stack = []
        /** @type {Map<string, Module>} */
        #modules = new Map()
        /** @type {Map<string, Module>} */
        #root_or_async_or_deferred_module = new Map()
        /** @type {number} */
        #first_request
        #freeze() {
            this.total_time = performance.now() - this.#first_request
            this.#stack.forEach((module) => this.leave(module.module_id))
            this.#finalized = true
            this.#async_compile_info.forEach((targets, from) => {
                const from_mod = this.#modules.get(String(from))
                if (!from_mod) return
                targets.forEach((x) => this.#modules.get(x)?.add_used_by(from_mod, 'async'))
            })
            this.#defer_compile_info.forEach((targets, from) => {
                const from_mod = this.#modules.get(String(from))
                if (!from_mod) return
                targets.forEach((x) => this.#modules.get(x)?.add_used_by(from_mod, 'defer'))
            })
            this.#merge_npm_nodes()

            this.modules = module_set_to_array(this.#modules)
            this.root_or_async_or_deferred_module = module_set_to_array(this.#root_or_async_or_deferred_module)

            delete Graph.prototype.connect
            delete Graph.prototype.enter
            delete Graph.prototype.leave
            delete Graph.prototype.set_compile_info
            this.modules.forEach((x) => x.freeze())
            delete Module.prototype.freeze
        }
        /** @type {Map<string, Set<string>>} */
        #async_compile_info = new Map()
        /** @type {Map<string, Set<string>>} */
        #defer_compile_info = new Map()
        /** @param {Array<[from: string | number, asyncTarget: string[], deferTarget: string[]]>} arr */
        set_compile_info(arr) {
            if (this.#finalized) return
            for (let [name, async, defer] of arr) {
                const n = (readable_name_cache[name] ??= simplified_name(String(name)))
                if (async.length) {
                    this.#async_compile_info.has(n) || this.#async_compile_info.set(n, new Set())
                    const s = this.#async_compile_info.get(n)
                    for (const x of async) s.add((readable_name_cache[x] ??= simplified_name(String(x))))
                }
                if (defer.length) {
                    this.#defer_compile_info.has(n) || this.#defer_compile_info.set(n, new Set())
                    const s = this.#defer_compile_info.get(n)
                    for (const x of async) s.add((readable_name_cache[x] ??= simplified_name(String(x))))
                }
            }
        }
        /**
         * @private
         * @param {string} dependency
         */
        connect(dependency) {
            if (this.#finalized) return
            this.#first_request ??= performance.now()
            dependency = readable_name_cache[dependency] ??= simplified_name(dependency)
            const dep = this.#modules.get(dependency)
            const current = this.#stack.at(-1)
            if (current) {
                dep.add_used_by(current)
            } else {
                this.#root_or_async_or_deferred_module.set(dep.module_id, dep)
            }
        }
        /**
         * @private
         * @param {string} current
         */
        enter(current) {
            if (this.#finalized) return
            this.#first_request ??= performance.now()
            current = readable_name_cache[current] ??= simplified_name(current)
            const new_mod = new Module(current)
            this.#modules.set(current, new_mod)
            this.connect(current)
            this.#stack.push(new_mod)
        }
        /**
         * @private
         * @param {string} current
         */
        leave(current) {
            if (this.#finalized) return
            current = readable_name_cache[current] ??= simplified_name(current)
            const last = this.#stack.at(-1)
            if (last?.module_id !== current) {
                throw new Error(`Ord? Stack ${this.#stack.map((x) => x.module_id).join('\n')}, out: ${current}`)
            }
            last.end_time = performance.now()
            this.#stack.pop()
            this.#stack.forEach((m) => (m.child_span_time += last.end_time - last.start_time - last.child_span_time))
        }

        #merge_npm_nodes() {
            // categorize all files by their package
            /** @type {Map<string, Set<Module>>} */
            const packages = new Map()
            for (const node of this.#modules.values()) {
                const npm_name = get_module_name(node.module_id)
                if (!npm_name) continue
                if (!packages.has(npm_name)) packages.set(npm_name, new Set())
                packages.get(npm_name).add(node)
            }
            // merge internal files of npm packages
            for (const [npm_name, files] of packages) {
                // no internal file
                if (files.size === 1) continue

                /** @type {Set<Module>} */
                const entry_files = new Set()
                /** @type {Set<Module>} */
                const normal_files = new Set()
                // if a file is imported by another file outside of the module directly, it's an entry file
                for (const file of files) {
                    for (const used_by of file.used_by) {
                        if (npm_name !== get_module_name(used_by.module_id)) {
                            entry_files.add(file)
                            break
                        }
                    }
                    entry_files.has(file) || normal_files.add(file)
                }
                if (normal_files.size === 0 || entry_files.size === 0) continue
                // if there is only one entry file, we can merge all files into this one
                if (entry_files.size === 1) {
                    const [entry] = entry_files.values()
                    entry.concat_modules = new Set()
                    for (const file of normal_files) {
                        entry.concat_with(file)
                        this.#modules.delete(file.module_id)
                    }
                } else {
                    // otherwise do a DFS to search all path into this file.
                    // only merge a file when there is only 1 [dominator](https://en.wikipedia.org/wiki/Dominator_(graph_theory)) entry file.
                    for (const file of normal_files) {
                        /** @type {Set<Module>} */
                        const entry_in_tree = new Set()
                        dfs(
                            file,
                            (m) => {
                                if (m === file) return
                                if (entry_files.has(m)) {
                                    entry_in_tree.add(m)
                                    return ['continue']
                                }
                                if (entry_files.size >= 2) return ['break']
                            },
                            'used_by',
                        )
                        if (entry_in_tree.size === 1) {
                            const [entry] = entry_in_tree
                            entry.concat_with(file)
                        }
                    }
                }
            }
        }
    }
    class Module {
        module_id
        constructor(/** @type {string} */ module_id) {
            this.module_id = module_id
        }
        /** @type {Set<Module>} */
        #used_by = new Set()
        /** @type {readonly Module[]} */
        get used_by() {
            return Object.freeze([...this.#used_by.values()])
        }
        /** @type {Set<Module>} */
        #imports = new Set()
        /** @type {readonly Module[]} */
        get imports() {
            return Object.freeze([...this.#imports.values()])
        }
        /** @type {Set<Module>} */
        #defer_used_by = new Set()
        /** @type {readonly Module[]} */
        get defer_used_by() {
            return Object.freeze([...this.#defer_used_by.values()])
        }
        /** @type {Set<Module>} */
        #async_used_by = new Set()
        /** @type {readonly Module[]} */
        get async_used_by() {
            return Object.freeze([...this.#async_used_by.values()])
        }
        /** @type {Set<Module>} */
        #defer_imports = new Set()
        /** @type {readonly Module[]} */
        get defer_imports() {
            return Object.freeze([...this.#defer_imports.values()])
        }
        /** @type {Set<Module>} */
        #async_imports = new Set()
        /** @type {readonly Module[]} */
        get async_imports() {
            return Object.freeze([...this.#async_imports.values()])
        }
        /**
         * Modules that this module depends
         * @type {Set<string> | undefined}
         */
        concat_modules
        start_time = performance.now()
        end_time = NaN
        child_span_time = 0
        get self_time() {
            return this.end_time - this.start_time - this.child_span_time
        }
        add_used_by(/** @type {Module} */ module, /** @type {'normal' | 'async' | 'defer'} */ type = 'normal') {
            if (this === module) return
            if (type === 'normal') {
                this.#used_by.add(module)
                module.#imports.add(this)
            } else if (type === 'async') {
                this.#async_used_by.add(module)
                module.#async_imports.add(this)
            } else if (type === 'defer') {
                this.#defer_used_by.add(module)
                module.#defer_imports.add(this)
            }
        }
        #remove_used_by(/** @type {Module} */ module) {
            this.#used_by.delete(module)
            module.#imports.delete(this)
        }
        concat_with(/** @type {Module} */ removing_node) {
            if (removing_node === this) return
            this.concat_modules ??= new Set()
            this.concat_modules.add(removing_node.module_id)

            const imports = [...removing_node.#imports]
            const used_by = [...removing_node.#used_by]
            imports.forEach((m) => {
                m.#remove_used_by(removing_node)
                m.add_used_by(this)
            })
            used_by.forEach((m) => {
                removing_node.#remove_used_by(m)
                this.add_used_by(m)
            })
        }
        freeze() {
            Object.defineProperties(this, {
                start_time: { value: this.start_time },
                end_time: { value: this.end_time },
                self_time: { value: this.self_time },
                used_by: { value: this.used_by },
                imports: { value: this.imports },
                async_used_by: { value: module_set_to_array(this.#async_used_by) },
                async_imports: { value: module_set_to_array(this.#async_imports) },
                defer_used_by: { value: module_set_to_array(this.#defer_used_by) },
                defer_imports: { value: module_set_to_array(this.#defer_imports) },
            })
        }
    }
    function of(object, config) {
        return ['object', { object, config }]
    }
    function line(...object) {
        return ['div', {}].concat(object)
    }
    globalThis.devtoolsFormatters = [
        ...(globalThis.devtoolsFormatters || []),
        {
            header: function (m) {
                if (!(m instanceof Module)) return null
                const self_time = parseInt(m.self_time.toFixed(2))
                const total_time = parseInt((m.end_time - m.start_time).toFixed(2))

                const self_fast = self_time < 4
                const result = line(['span', self_fast ? { style: FAST_TIME_STYLE } : {}, m.module_id])

                if (self_time !== 0 || self_time !== total_time)
                    result.push(['span', { style: get_time_style(self_time) }, ' ', self_time])
                if (self_time !== total_time)
                    result.push(
                        ['span', { style: FAST_TIME_STYLE }, '/'],
                        ['span', { style: get_time_style(total_time) }, total_time, 'ms'],
                    )
                else if (self_time !== 0) result.push(['span', { style: get_time_style(self_time) }, 'ms'])

                if (m.imports.length || m.used_by.length) result.push(' (')
                if (m.used_by.length) result.push(['span', { style: USED_BY_STYLE }, `by ${m.used_by.length}`])
                if (m.imports.length && m.used_by.length) result.push(' ')
                if (m.imports.length) result.push(['span', { style: IMPORT_STYLE }, `to ${m.imports.length}`])
                if (m.imports.length || m.used_by.length) result.push(')')

                if (m.concat_modules?.size)
                    result.push(['span', { style: MERGED_MODULE_STYLE }, ` +${m.concat_modules.size} modules`])
                return result
            },
            hasBody: (m) =>
                m instanceof Module &&
                (m.used_by.length ||
                    m.imports.length ||
                    m.async_imports.length ||
                    m.async_used_by.length ||
                    m.defer_imports.length ||
                    m.defer_used_by.length),
            body: function (m, config) {
                if (!(m instanceof Module)) return null
                const level = config?.level ?? 1
                const next_config = { level: level + 1 }
                const import_style = { style: `margin-left: ${level * 1}em;${IMPORT_STYLE}` }
                const used_by_style = { style: `margin-left: ${level * 1}em;${USED_BY_STYLE}` }
                return [
                    'div',
                    {},
                    m.used_by.length && [
                        'div',
                        used_by_style,
                        'used by',
                        m.used_by.length >= 10
                            ? line(of(m.used_by, next_config))
                            : line(...m.used_by.map((x) => line(of(x)))),
                    ],
                    m.imports.length && [
                        'div',
                        import_style,
                        'imports',
                        m.imports.length >= 10
                            ? line(of(m.imports, next_config))
                            : line(...m.imports.map((x) => line(of(x)))),
                    ],
                    m.async_used_by.length && ['div', used_by_style, 'async used by', line(of(m.async_used_by))],
                    m.defer_used_by.length && ['div', used_by_style, 'defer used by', line(of(m.defer_used_by))],
                    m.async_imports.length && ['div', import_style, 'async imports', line(of(m.async_imports))],
                    m.defer_imports.length && ['div', import_style, 'defer imports', line(of(m.defer_imports))],
                ].filter(Boolean)
            },
        },
    ]
    Object.setPrototypeOf(Graph.prototype, null)
    Object.setPrototypeOf(Module.prototype, null)
    globalThis.measure = new Graph()

    function module_set_to_array(/** @type {Set<Module> | Map<any, Module>} */ set) {
        if (set.size === 0) return EMPTY
        const arr = [...set.values()]
        Object.assign(arr, Object.fromEntries(set instanceof Map ? set : arr.map((x) => [x.module_id, x])))
        Object.freeze(arr)
        return arr
    }

    /**
     * @template T
     * @param {Module} module
     * @param {(module: Module) => void | ['completion', T] | ['continue'] | ['break']} callback
     * @param {'all' | 'imports' | 'used_by'} direction
     * @returns {T | undefined}
     */
    function dfs(module, callback, direction = 'all') {
        const visited = new Set()
        const op = inner(module)
        if (op?.[0] === 'completion') return op[1]
        return undefined

        function inner(/** @type {Module} */ module) {
            visited.add(module)
            const op = callback(module)
            if (op) {
                if (op[0] === 'continue') return
                return op
            }
            if (direction !== 'used_by') {
                for (const node of module.imports) {
                    if (visited.has(node)) continue
                    inner(node)
                }
            }
            if (direction !== 'imports') {
                for (const node of module.used_by) {
                    if (visited.has(node)) continue
                    inner(node)
                }
            }
        }
    }
    /**
     * @param {string} module
     * @returns {string}
     */
    function get_module_name(module) {
        return module.startsWith('@')
            ? module.slice(0, module.indexOf('/', module.indexOf('/')))
            : module.slice(0, module.indexOf('/'))
    }
    /**
     * @param {string} module
     */
    function simplified_name(module) {
        const npm_name = (npm_name_cache[module] ??= module.match(regex)?.groups?.key)
        if (npm_name) {
            const no_version = npm_name.startsWith('@')
                ? '@' + npm_name.split('@')[1].replace('+', '/')
                : npm_name.split('@')[0]
            return module.replace(new RegExp(`.+/node_modules/${no_version}`), no_version)
        } else {
            return module
        }
    }
    /**
     * @param {number} time
     */
    function get_time_style(time) {
        return time > 10 ? SLOW_TIME_STYLE : time < 4 ? FAST_TIME_STYLE : TIME_STYLE
    }
    undefined
})()
