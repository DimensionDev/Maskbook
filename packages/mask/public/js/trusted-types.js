/// <reference path="../../polyfills/types/dom.d.ts" />
if (typeof trustedTypes === 'object' && location.protocol.includes('extension')) {
    // https://github.com/TanStack/query/issues/5417
    const ReactQueryStyle = `
            .ReactQueryDevtoolsPanel * {
              scrollbar-color: #132337 #3f4e60;
            }

            .ReactQueryDevtoolsPanel *::-webkit-scrollbar, .ReactQueryDevtoolsPanel scrollbar {
              width: 1em;
              height: 1em;
            }

            .ReactQueryDevtoolsPanel *::-webkit-scrollbar-track, .ReactQueryDevtoolsPanel scrollbar-track {
              background: #132337;
            }

            .ReactQueryDevtoolsPanel *::-webkit-scrollbar-thumb, .ReactQueryDevtoolsPanel scrollbar-thumb {
              background: #3f4e60;
              border-radius: .5em;
              border: 3px solid #132337;
            }
          `
    trustedTypes.createPolicy('default', {
        // do not add createHTML or createScript.
        // createScriptURL is safe because according to the CSP we have, it is impossible to
        // include/create a script from cross-origin.
        createScriptURL: (string) => string,

        createHTML: (html) => {
            if (html !== ReactQueryStyle) {
                throw new TypeError('Failed to convent a String to a TrustedHTML')
            }
            return html
        },
    })
}

undefined
