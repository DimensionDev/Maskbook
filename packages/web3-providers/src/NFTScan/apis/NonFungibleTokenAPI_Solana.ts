import { EMPTY_LIST } from '@masknet/shared-base'
import {
    createIndicator,
    createNextIndicator,
    createPageable,
    HubIndicator,
    HubOptions,
    NonFungibleCollection,
    NonFungibleTokenContract,
    NonFungibleTokenEvent,
    Pageable,
} from '@masknet/web3-shared-base'
import { ChainId, SchemaType, isValidChainId } from '@masknet/web3-shared-solana'
import urlcat from 'urlcat'
import type { NonFungibleTokenAPI } from '../../types/index.js'
import {
    createNonFungibleAsset,
    createNonFungibleCollection,
    createNonFungibleTokenContract,
    createNonFungibleTokenEvent,
    fetchFromNFTScanV2,
} from '../helpers/Solana.js'
import type { PageableResponse, Response, Solana } from '../types/index.js'

export class NFTScanNonFungibleTokenAPI_Solana implements NonFungibleTokenAPI.Provider<ChainId, SchemaType> {
    async getAsset(address: string, _?: string, { chainId = ChainId.Mainnet }: HubOptions<ChainId> = {}) {
        if (!isValidChainId(chainId)) return
        const path = urlcat('/api/sol/assets/:address', {
            address,
            show_attribute: true,
        })
        const response = await fetchFromNFTScanV2<Response<Solana.Asset>>(chainId, path)
        if (!response?.data) return
        const asset = createNonFungibleAsset(chainId, response.data)
        return asset
    }

    async getAssets(account: string, { chainId = ChainId.Mainnet, indicator, size = 20 }: HubOptions<ChainId> = {}) {
        if (!isValidChainId(chainId)) return createPageable(EMPTY_LIST, createIndicator(indicator))
        const path = urlcat('/api/sol/account/own/:from', {
            from: account,
            account_address: account,
            cursor: indicator?.id,
            limit: size,
            show_attribute: true,
        })
        const response = await fetchFromNFTScanV2<PageableResponse<Solana.Asset>>(chainId, path)
        const assets = response?.data.content.map((x) => createNonFungibleAsset(chainId, x)) ?? EMPTY_LIST
        return createPageable(
            assets,
            createIndicator(indicator),
            response?.data.next ? createNextIndicator(indicator, response?.data.next) : undefined,
        )
    }

    async getAssetsByCollection(
        name: string,
        { chainId = ChainId.Mainnet, indicator, size = 20 }: HubOptions<ChainId> = {},
    ) {
        if (!isValidChainId(chainId)) return createPageable(EMPTY_LIST, createIndicator(indicator))
        const path = urlcat('/api/sol/assets/collection/:name', {
            name,
            collection: name,
            cursor: indicator?.id,
            limit: size,
        })
        const response = await fetchFromNFTScanV2<PageableResponse<Solana.Asset>>(chainId, path)
        const assets = response?.data?.content.map((x) => createNonFungibleAsset(chainId, x)) ?? EMPTY_LIST
        return createPageable(
            assets,
            createIndicator(indicator),
            response?.data.next ? createNextIndicator(indicator, response?.data.next) : undefined,
        )
    }

    async getCollectionsByKeyword(
        keyword: string,
        { chainId = ChainId.Mainnet, indicator, size = 20 }: HubOptions<ChainId, HubIndicator> = {},
    ): Promise<Pageable<NonFungibleCollection<ChainId, SchemaType>, HubIndicator>> {
        if (!isValidChainId(chainId)) return createPageable(EMPTY_LIST, createIndicator(indicator))
        const path = '/api/sol/collections/filters'
        const response = await fetchFromNFTScanV2<Response<Solana.Collection[]>>(chainId, path, {
            method: 'POST',
            body: JSON.stringify({
                symbol: keyword,
                offset: (indicator?.index ?? 0) * size,
                limit: size.toString(),
            }),
        })
        const collections = response?.data.map((x) => createNonFungibleCollection(chainId, x)) ?? EMPTY_LIST
        return createPageable(collections, createIndicator(indicator))
    }

    async getCollection(
        name: string,
        { chainId = ChainId.Mainnet }: HubOptions<ChainId, HubIndicator> = {},
    ): Promise<NonFungibleCollection<ChainId, SchemaType> | undefined> {
        const path = urlcat('/api/sol/collections/:name', {
            collection: name,
        })
        const response = await fetchFromNFTScanV2<Response<Solana.Collection>>(chainId, path)
        if (!response?.data) return
        return createNonFungibleCollection(chainId, response.data)
    }

    async getContract(
        address: string,
        { chainId = ChainId.Mainnet }: HubOptions<ChainId, HubIndicator> = {},
    ): Promise<NonFungibleTokenContract<ChainId, SchemaType> | undefined> {
        const path = urlcat('/api/v2/collections/:address', {
            address,
        })
        const response = await fetchFromNFTScanV2<Response<Solana.Collection>>(chainId, path)
        if (!response?.data) return
        return createNonFungibleTokenContract(chainId, response.data)
    }

    async getEvents(
        address: string,
        _: string,
        { chainId = ChainId.Mainnet, indicator, size = 20 }: HubOptions<ChainId, HubIndicator> = {},
    ): Promise<Pageable<NonFungibleTokenEvent<ChainId, SchemaType>>> {
        if (!isValidChainId(chainId)) return createPageable(EMPTY_LIST, createIndicator(indicator))
        const path = urlcat('/api/sol/transactions/:address', {
            address,
            token_address: address,
            limit: size,
            cursor: indicator?.id,
        })
        const response = await fetchFromNFTScanV2<PageableResponse<Solana.Transaction>>(chainId, path)
        const events = response?.data.content.map((x) => createNonFungibleTokenEvent(chainId, x)) ?? EMPTY_LIST
        return createPageable(
            events,
            createIndicator(indicator),
            response?.data.next ? createNextIndicator(indicator, response?.data.next) : undefined,
        )
    }
}
