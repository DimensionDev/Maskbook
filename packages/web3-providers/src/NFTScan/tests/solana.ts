import { describe, expect, it } from 'vitest'
import { NFTScanNonFungibleTokenAPI_Solana } from '../apis/NonFungibleTokenAPI_Solana.js'

describe('NFTScan Solana', () => {
    const provider = new NFTScanNonFungibleTokenAPI_Solana()
    it.skipIf(!process.env.RUN_SKIP_TESTS)(
        'should get asset correctly',
        async () => {
            // cspell:ignore 9quqwD7X3ywS2QbibHGj8i95UAr6ypMDhXiv5ZmwKSvY
            const asset = await provider.getAsset('9quqwD7X3ywS2QbibHGj8i95UAr6ypMDhXiv5ZmwKSvY')
            expect(asset!.traits?.sort((a, z) => a.type.charCodeAt(0) - z.type.charCodeAt(0))).toMatchSnapshot()
        },
        { retry: 3 },
    )

    it.skipIf(!process.env.RUN_SKIP_TESTS)(
        'show get assets by account correctly',
        async () => {
            // cspell:ignore 4Qzi1RHo3gnQ4oQrq3UccHehqA7nKo7DpMbnZUy1zpHG
            const list = await provider.getAssets('4Qzi1RHo3gnQ4oQrq3UccHehqA7nKo7DpMbnZUy1zpHG', { size: 1 })
            expect(list.data[0]!.traits?.sort((a, z) => a.type.charCodeAt(0) - z.type.charCodeAt(0))).toMatchSnapshot()
        },
        { retry: 3 },
    )
})
