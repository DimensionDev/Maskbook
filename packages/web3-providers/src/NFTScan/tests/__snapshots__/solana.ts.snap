// Vitest Snapshot v1

exports[`NFTScan Solana > should get asset correctly 1`] = `
[
  {
    "rarity": "12.61%",
    "type": "Background",
    "value": "Grey",
  },
  {
    "rarity": "3.83%",
    "type": "Clothes",
    "value": "Black Tee",
  },
  {
    "rarity": "5.07%",
    "type": "Eyes",
    "value": "Sad",
  },
  {
    "rarity": "2.82%",
    "type": "Eyewear",
    "value": "Circle Sunglasses",
  },
  {
    "rarity": "2.01%",
    "type": "Fur",
    "value": "Wood",
  },
  {
    "rarity": "2.12%",
    "type": "Hat",
    "value": "Backwards Skate Cap",
  },
  {
    "rarity": "9.22%",
    "type": "Mouth",
    "value": "Wry",
  },
]
`;

exports[`NFTScan Solana > show get assets by account correctly 1`] = `
[
  {
    "rarity": "12.6%",
    "type": "Background",
    "value": "Green",
  },
  {
    "rarity": "0.88%",
    "type": "Clothes",
    "value": "Tweed Suit",
  },
  {
    "rarity": "9.4%",
    "type": "Eyes",
    "value": "Disappointed",
  },
  {
    "rarity": "1.96%",
    "type": "Fur",
    "value": "Lofi",
  },
  {
    "rarity": "5.51%",
    "type": "Hat",
    "value": "Skate Cap",
  },
  {
    "rarity": "2.53%",
    "type": "Mouth",
    "value": "Raspberry",
  },
]
`;
